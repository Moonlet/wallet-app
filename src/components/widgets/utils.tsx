import React from 'react';
import { DataType, ICurrencyData, IData, IDataStyle, ITextData } from './types';
import { Text } from '../../library';
import BigNumber from 'bignumber.js';
import { formatNumber } from '../../core/utils/format-number';
import { normalize, normalizeFontAndLineHeight } from '../../styles/dimensions';

const formatTextData = (data: ITextData): string => {
    let text = '';
    // Static text
    text = String(data.value);

    if (data?.params) {
        // TODO
    }

    // Translated text
    // TODO

    return text;
};

const amountFromStd = (value: BigNumber | number | string, decimals: number): BigNumber => {
    return new BigNumber(value).dividedBy(new BigNumber(10).pow(decimals));
};

const formatCurrencyData = (data: ICurrencyData): string => {
    let text = '';
    let value = data.value;

    if (data?.decimals) {
        value = amountFromStd(value, data.decimals).toFixed();
    }

    text = formatNumber(Number(value), {
        currency: data.symbol,
        maximumFractionDigits: data?.round?.decimals
    });

    if (data?.beautify) {
        const out = beautify(value);
        text = new BigNumber(out.value).toFixed(data.beautify.decimals) + out.unit;
    }

    return text;
};

const formatStyles = (style: IDataStyle): { [key: string]: string | number } => {
    const finalStyle = {};

    for (const s of Object.keys(style)) {
        const sVal: any = style[s];

        if (typeof sVal === 'object') {
            if (sVal.fn === 'normalize') {
                finalStyle[s] = normalize(sVal.value);
            }
            if (sVal.fn === 'normalizeFontAndLineHeight') {
                finalStyle[s] = normalizeFontAndLineHeight(sVal.value);
            }
        } else {
            finalStyle[s] = style[s];
        }
    }

    return finalStyle;
};

const beautify = (
    value: BigNumber | string | number,
    fromValue?: number
): {
    value: BigNumber | string;
    unit: string;
} => {
    fromValue = fromValue || 1000;

    value = new BigNumber(value);

    let out = { value, unit: '' };

    if (value.gte(fromValue)) {
        const units = [
            [1000000000, 'B'],
            [1000000, 'M'],
            [1000, 'k']
        ];

        for (const unit of units) {
            if (value.gte(unit[0])) {
                out = {
                    value: value.div(unit[0]),
                    unit: unit[1].toString()
                };
                break;
            }
        }
    }

    return out;
};

export const formatDataJSXElements = (data: IData | IData[], baseStyle: any): JSX.Element[] => {
    if (!Array.isArray(data)) {
        data = [data];
    }

    const finalData: JSX.Element[] = [];

    for (const d of data) {
        switch (d.type) {
            case DataType.TEXT:
                finalData.push(
                    <Text style={[baseStyle, d?.style && formatStyles(d.style)]}>
                        {formatTextData(d.data as ITextData)}
                    </Text>
                );
                break;

            case DataType.CURRENCY:
                finalData.push(
                    <Text style={[baseStyle, d?.style && formatStyles(d.style)]}>
                        {formatCurrencyData(d.data as ICurrencyData)}
                    </Text>
                );
                break;

            default:
                break;
        }
    }

    return finalData;
};
