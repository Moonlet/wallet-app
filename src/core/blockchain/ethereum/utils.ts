import { TransactionStatus } from '../../wallet/types';

export const getEthStatusByCode = (status): TransactionStatus => {
    switch (parseInt(status, 16)) {
        case 0:
            return TransactionStatus.FAILED;
        case 1:
            return TransactionStatus.SUCCESS;
        case 2:
            return TransactionStatus.PENDING;
        default:
            return TransactionStatus.FAILED;
    }
};
