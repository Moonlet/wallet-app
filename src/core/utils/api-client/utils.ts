import moment from 'moment';
import CONFIG from '../../../config';
import * as schnorr from '@zilliqa-js/crypto/dist/schnorr';

// Refactor this using ntp time
export const getTimestamp = () => {
    return moment().unix() * 1000;
};

export const getDomain = () => {
    return new URL(CONFIG.walletApiBaseUrl).host;
};

export const getSignature = (
    data: any,
    walletPrivateKey: string,
    walletPublicKey: string
): string => {
    const sig = schnorr.sign(
        Buffer.from(JSON.stringify(data), 'hex'),
        Buffer.from(walletPrivateKey, 'hex'),
        Buffer.from(walletPublicKey, 'hex')
    );

    return sig.r.toString('hex') + sig.s.toString('hex');
};
