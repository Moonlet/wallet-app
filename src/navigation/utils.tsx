import React from 'react';
import { View, Text } from 'react-native';
import { Icon } from '../components/icon';
import { COLORS } from '../styles/colors';
import { StackViewStyleInterpolator } from 'react-navigation-stack';
import { normalize } from '../library';

export const DummyScreen = () => (
    <View style={{ display: 'flex', justifyContent: 'center', flex: 1 }}>
        <Text style={{ textAlign: 'center' }}>some screen</Text>
    </View>
);

// TODO use theme for colors
export const menuIcon = (icon: string) => ({ focused }: any) => (
    <Icon
        name={icon}
        size={normalize(25)}
        style={{ color: focused ? COLORS.AQUA : COLORS.DARK_GRAY }}
    />
);

// remove animation when transitioning to any of `noAnimationScreens`
export const removeAnimation = (noAnimationScreens: string[]) => sceneProps => {
    if (noAnimationScreens.indexOf(sceneProps.scene.route.routeName) !== -1) {
        return null;
    }
    return StackViewStyleInterpolator.forHorizontal(sceneProps);
};
