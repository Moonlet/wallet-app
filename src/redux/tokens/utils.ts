import { ITokenState } from '../wallets/state';

export const accountToken = (
    symbolKey: string,
    order: number,
    options?: { active?: boolean }
): ITokenState => {
    return {
        symbol: symbolKey,
        order,
        active: options?.active !== undefined ? options.active : true,
        balance: {
            value: '0',
            inProgress: false,
            timestamp: undefined,
            error: undefined,
            available: '0',
            total: '0',
            detailed: {}
        }
    };
};
