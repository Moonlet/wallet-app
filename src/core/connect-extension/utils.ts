import { getBlockchain } from '../blockchain/blockchain-factory';
import { NavigationParams } from 'react-navigation';
import { NavigationService } from '../../navigation/navigation-service';

export const getUrlParams = (search = ``) => {
    const hashes = search.slice(search.indexOf('?') + 1).split('&');
    return hashes.reduce((params, hash) => {
        const split = hash.indexOf('=');

        if (split < 0) {
            return { ...params, [hash]: null };
        }

        const key = hash.slice(0, split);
        const val = hash.slice(split + 1);

        return { ...params, [key]: decodeURIComponent(val) };
    }, {});
};

export const signExtensionTransaction = payload => {
    const data = payload.params[0];
    const blockchainConfig = getBlockchain(data.selectedAccount.blockchain).config;
    const navigationParams: NavigationParams = {
        blockchain: data.selectedAccount.blockchain,
        accountIndex: data.selectedAccount.index,
        token: data.token,
        tokenLogo: blockchainConfig.tokens[blockchainConfig.coin].icon,
        extensionTransactionPayload: payload
    };

    if (NavigationService.getCurrentRoute() === 'Token') {
        NavigationService.replace('Token', navigationParams);
    } else {
        NavigationService.navigate('Token', navigationParams);
    }
};
