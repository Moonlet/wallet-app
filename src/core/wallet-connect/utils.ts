import { NavigationService } from '../../navigation/navigation-service';
import { getBlockchain } from '../blockchain/blockchain-factory';
import { NavigationParams } from 'react-navigation';

export const signExtensionTransaction = payload => {
    const data = payload.params[0];
    const blockchainConfig = getBlockchain(data.selectedAccount.blockchain).config;
    const navigationParams: NavigationParams = {
        blockchain: data.selectedAccount.blockchain,
        accountIndex: data.selectedAccount.index,
        token: data.token,
        tokenLogo: blockchainConfig.tokens[blockchainConfig.coin].icon,
        extensionTransactionPayload: payload
    };

    if (NavigationService.getCurrentRoute() === 'Token') {
        NavigationService.replace('Token', navigationParams);
    } else {
        NavigationService.navigate('Token', navigationParams);
    }
};
