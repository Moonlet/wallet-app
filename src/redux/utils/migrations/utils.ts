import { accountToken } from '../../tokens/utils';
import { Blockchain } from '../../../core/blockchain/types';
import { IWalletState } from '../../wallets/state';

export const addToken = (
    state: any,
    token: any,
    data: {
        blockchain: Blockchain;
        chainId: string;
        tokenActive: boolean;
        order?: number;
    }
) => {
    if (
        state.tokens[data.blockchain] &&
        state.tokens[data.blockchain][data.chainId] &&
        state.tokens[data.blockchain][data.chainId][token.symbol]
    ) {
        // Update toke contract address
        state.tokens[data.blockchain][data.chainId][token.symbol].contractAddress =
            token.contractAddress;
    } else {
        // Add token Token
        state.tokens = {
            ...state.tokens,
            [data.blockchain]: {
                ...(state.tokens && state.tokens[data.blockchain]),
                [data.chainId]: {
                    ...(state.tokens &&
                        state.tokens[data.blockchain] &&
                        state.tokens[data.blockchain][data.chainId]),
                    [token.symbol]: token
                }
            }
        };
    }

    // Add tokens on accounts
    for (const wallet of Object.values(state.wallets)) {
        for (const account of (wallet as IWalletState).accounts) {
            if (account.blockchain === data.blockchain) {
                if (account.tokens[data.chainId] && account.tokens[data.chainId][token.symbol]) {
                    // Token has been already added
                } else {
                    // Add Token
                    account.tokens[data.chainId][token.symbol] = accountToken(
                        token.symbol,
                        data.order || 999,
                        {
                            active: data.tokenActive
                        }
                    );
                }
            }
        }
    }

    return {
        ...state
    };
};
