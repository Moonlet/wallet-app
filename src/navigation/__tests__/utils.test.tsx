import React from 'react';
import { shallow } from 'enzyme';
import { DummyScreen, menuIcon } from '../utils';

export default describe('NavigationUtils', () => {
    describe('DummyScreen', () => {
        test('renders corectly', () => {
            expect(shallow(<DummyScreen />).debug()).toMatchSnapshot();
        });
    });

    describe('menuIcon', () => {
        test('renders corectly', () => {
            const Comp = menuIcon('accounting-coins-stack');
            expect(shallow(<Comp />).debug()).toMatchSnapshot('unfocused');
            expect(shallow(<Comp focused={true} />).debug()).toMatchSnapshot('focused');
        });
    });
});
