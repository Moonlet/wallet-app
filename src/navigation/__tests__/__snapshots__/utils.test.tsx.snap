// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationUtils DummyScreen renders corectly 1`] = `
"<View style={{...}}>
  <Text style={{...}}>
    some screen
  </Text>
</View>"
`;

exports[`NavigationUtils menuIcon renders corectly: focused 1`] = `"<Icon name=\\"accounting-coins-stack\\" size={25} style={{...}} allowFontScaling={false} />"`;

exports[`NavigationUtils menuIcon renders corectly: unfocused 1`] = `"<Icon name=\\"accounting-coins-stack\\" size={25} style={{...}} allowFontScaling={false} />"`;
